doc/epic.1
.Dd April 22, 1999
.Dt EPIC 1
.Sh NAME
.Nm epic
.Nd Internet Relay Chat client for UNIX like systems
.Sh SYNOPSIS
.Nm epic
.Op Ar \-a
.Op Ar \-b
.Op Ar \-B
.Op Ar \-c chan
.Op Ar \-d 
.Op Ar \-f 
.Op Ar \-F 
.Op Ar \-h 
.Op Ar \-H hostname
.Op Ar \-l filename
.Op Ar \-L filename 
.Op Ar \-n nickname
.Op Ar \-o
.Op Ar \-O
.Op Ar \-p port
.Op Ar \-q 
.Op Ar \-v
.Op Ar \-x
.Op Ar \-z username
.Op Ar nickname
.Op Ar server description list
.Sh DESCRIPTION
The
.Nm ircII/EPIC
program is a unix-based character oriented user agent ('client') to 
Internet Relay Chat.
It is a fully functional ircII client with many useful extensions.  
This version works with all modern irc server classes as of early 1999.
.Pp
.Sh OPTIONS
.Bl -tag -width flag
.It Fl a
Append the server description list to the default server list.
The default behavior is for the server description list to replace the default server list.
.It Fl b
Operate in so called
.Dq bot mode.  
This implies the 
.Op Ar \-d
option.  
EPIC will 
.Xr fork 2
immediately and the parent process will exit, returning you to your shell.
Some system administrators do not look kindly to their users running bots, and they have disabled this option.  
Even if your administrator has not disabled it, you should not assume this gives you automatic permission to run a bot.  
If you do run a bot without permission, your administrator may get very angry with you, and possibly revoke your account.  
In addition, most IRC operators on public irc networks have very little tolerance for people who run bots.
So just a word of caution, make sure that your system administrator and your irc administrator have given you permission before you run a bot.
.It Fl B
Force the startup file to be loaded immediately rather than waiting until a connection to a server is established.
.It Fl c Ar chan
Join the specified channel the first time you successfully connect to a server.
.It Fl d
Operate in 
.Dq dumb mode.
The client will not put up a full screen display, and will read from standard input and write to standard output.  
This is useful if the output normally looks awful (because you are using an incorrect
.Ev TERM 
setting, or your terminal description is spectacularly broken), or you just don't want to use the pretty interface.  
This option will be turned on automatically if your current 
.Ev TERM
setting is not capable of a full screen display.
.It Fl f
Force use of hardware flow control.  
With this option, the control-S and control-Q keys are probably not available to be bound to something else.
.It Fl F
Disable use of hardware flow control.  
With this option, the control-S and control-Q keys are available to be bound to something else.
However, you will not have hardware flow control.
.It Fl h
Display a moderately concise help message and exit immediately.
.It Fl H Ar hostname
Use the IP address of the specified hostname as your default IP address.
This can be used if you have multiple IP addresses on the same machine and you want to use an address other than the default address.  
You might need to use this option when
.Xr gethostname 3
does not return a hostname (in some poorly configured NIS environments).
The use of multiple IP addresses on a single machine is commonly referred to as "virtual hosting", and each IP address is a "virtual host".   
Please understand that an irc client may not tell the irc server what your hostname should be:  the server alone determines that.  
Servers typically use the canonical hostname for an IP address as your hostname.
Because of this, this option will not permit you to use a CNAME (secondary hostname for an IP address), because the server will use the canonical hostname instead.
This option overrides the 
.Ev IRCHOST 
environment variable.
.It Fl l Ar filename,[filename]
Use the specified filename(s) as the startup file.  
The startup file is loaded the first time you successfully connect to a server, unless you specify the
.Op Ar \-B
option.  
This overrides the
.Ev IRCRC
environment variable.  
If this option is not specified, and the 
.Ev IRCRC
environment variable is not set, then 
.Pa ~/.ircrc
is the default startup file.
.It Fl n Ar nickname
Use the specified nickname as the default nickname whenever you connect to an irc server.  
This option overrides the
.Ev IRCNICK
environment variable.  
This option can be overridden if you specify nickname argument in the command line (see below).
.It Fl o
Force use of IEXTEN termios characters.  
POSIX systems are allowed to reserve additional control characters to perform special actions when IEXTEN is turned on.  
On 4.4BSD, the control-V and control-O keys are used by IEXTEN and thus cannot be used in key bindings within EPIC since the terminal never sends them to EPIC.
.It Fl O
Disable use of IEXTEN termios characters.  
This makes all of the keys reserved by your system's IEXTEN termios option available to be used in key bindings.  
On 4.4BSD, this flag is necessary if you want to use control-V and control-O in your key bindings.
.It Fl p Ar port
Use the specified port as the default port for new server connections.
The default port is usually 6667.  
Make sure that the servers you want to connect to are listening on this port before you try to connect there.
.It Fl q
Suppress the loading of any file when you first establish a connection to an irc server.
.It Fl v
Output version identification (VID) information and exit.
.It Fl x
This undocumented feature turns on all of the XDEBUG flags.  
Refer to the help files for XDEBUG if you want to know what happens if you use this.
.It Fl z Ar username
Use the specified username when negotiating a connection to a new irc server.
This overrides the 
.Ev IRCUSER
environment variable.  
If this option is not specified, then the user name specified in 
.Pa /etc/passwd
for your user is used.  
This feature was formerly undocumented, but with the rise and popularity and use of
.Xr identd 8
this option is much less useful than it once was.  
Requests to have this option removed will probably be ignored.  
If you don't want your users to spoof their usernames, install identd, and do everyone on IRC a favor.
.It nickname
The first bare word found is taken as the default nickname to use.  
This overrides all other options, including the -n option and the 
.Ev IRCNICK 
environment variable.  
If all else fails, then the client uses your login name as the default nickname.
.It server,[server]
After the nickname, a list of one or more server specifications can be listed.
Unless you specify the -a option, this will replace your default server list!
The -a option forces any servers listed here to be appended to the default server list.
The format for server specifications is:
.sp
.Dl hostname:port:password:nick
.sp
Any item can be omitted by leaving the field blank, and any trailing colons can also be omitted.
.Sh "DETAILED DESCRIPTION"
.Ss "The Screen:"
The screen is split into two parts, separated by an inverse-video status line (if supported).
The upper (larger) part of the screen displays responses from the
.Xr ircd 8
server.
The lower part of the screen (a single line) accepts keyboard input.
.sp
Some terminals do not support certain features required by
.Nm epic
, in which case you receive a message stating this.
If this occurs, try changing the terminal type or run
.Nm epic
with the -d option.
.Ss "Irc Commands:"
Any line beginning with the slash character 
.Dq /
is regarded as an
.Nm epic
command (the command character may be changed).  
Any line not beginning with this character is treated as a message to be sent to the current channel.
The client has a built in help system.  
Install the help files (they should be available at the same place you got the client) and then type
.Dq /help
to open up the help system.
.Ss "The .ircrc File:"
When
.Nm epic
is executed, it checks the user's home directory for a
.Pa ~/.ircrc 
file, executing the commands in the file. 
Commands in this file do not need to have a leading slash character 
.Dq /
This allows predefinition of aliases and other features.
.Sh PRACTICAL EXAMPLES
Certainly any description of 
.Nm epic 
in this man page will be sorely inadequate because most of the confusion doesn't even start until after you get the client to connect to a server.  
But if you really have problems getting the client to connect to a server, try some of these:
.Bl -tag -width "epic"
.It Nm epic
Try this first.  
This will assume all the defaults.  
If the person who is maintaining epic at your site has done a halfway decent job, this will put you on a server that is somewhat local to you. 
.It Nm "epic nickname irc.domain.com"
or something similar will attempt to connect to the irc server running on the host "irc.domain.com" (fill in a real irc server here) with the nickname of well, "nickname".  
This is the most common way to specify an alternate server to use.
.It Nm "epic nickname irc.domain.com:6664"
Sometimes, some servers are really busy, and it can take them a long time to establish a connection with you on the default port (6667).  

script/history
if (word(2 $loadinfo()) != 'pf') {
	load -pf $word(1 $loadinfo());
	return;
};

#	Copyright (c) 2007 EPIC Software Labs
#	Written by BlackJac@EFNet
#
#	Version: 1.2.2007.09.28.1
#
#	This script simulates the history feature of epic4 for epic5. The
#	behavior and key bindings should be nearly identical to the original.
#	Several significant improvements have been made over the original epic4
#	implementation.
#
#	Settings:
#		set history [0|<positive integer>]
#			Controls the maximum number of entries to be stored in
#			the client's history buffer. Setting it to 0 purges the
#			history buffer and disables command history.
#		set history_circleq [on|off|toggle]
#			Controls the behavior of history browsing when you reach
#			the beginning or end of the buffer. If set to on, when
#			you reach the beginning of the buffer and press the
#			backward_history keybinding (defaulted to the up arrow)
#			again, the history will circle back to the end of the
#			buffer. In addition, when you reach the end of the
#			buffer and press the forward_history keybinding
#			(defaulted to the down arrow) again, the history will
#			circle back to the beginning of the buffer.
#		set history_persistent [on|off|toggle]
#			Controls whether the current session's history buffer
#			will be saved on exit and read back into memory the next
#			time the script is loaded.
#		set history_remove_dupes [on|off|toggle]
#			Controls whether history buffer entries that are exact
#			duplicates of the most recent entry are removed from the
#			buffer.
#		set history_save_file <filename>
#			Default filename to use when saving the current history
#			buffer or reading a previous buffer and adding it to the
#			current buffer.
#		set history_save_position [on|off|toggle]
#			Controls whether the backward_history or forward_history
#			keybindings will start scrolling through the history
#			buffer from the spot where a previous
#			/!<indexnum|pattern> match was found, or from the most
#			recent history buffer entry.
#		set history_timestamp [on|off|toggle]
#			Controls whether /history displays the timestamp of each
#			command.
#
#	Commands:
#		/history [<indexnum>]
#			Returns a list of all commands in the history buffer, or
#			all commands up to <indexnum>.
#		/!<indexnum|pattern>
#			Retrieves history buffer entries and outputs them to the
#			input line. Use /!<indexnum> to retrieve <indexnum> or
#			/!<pattern> to retrieve the most recent entry matching
#			<pattern>.
#
#	Functions:
#		$historyctl(add <text>)
#			Adds <text> to the next available index number in the
#			history buffer. 
#			Returns 0 if the history buffer is disabled.
#			Returns 1 if successful.
#		$historyctl(delete <indexnum>)
#			Removes <indexnum> from the history buffer.
#			Returns -2 if the history buffer is disabled.
#			Returns -1 if <indexnum> does not exist.
#			Returns 0 if successful.
#		$historyctl(get <indexnum>)
#			Returns the history buffer item <indexnum> if it exists,
#			or nothing.
#		$historyctl(index <indexnum>)
#			Moves the history buffer scrollback pointer to item
#			<indexnum>.
#			Returns 0 if <indexnum> does not exist.
#			Returns 1 if successful.
#		$historyctl(read [<filename>])
#			Reads into the history buffer the contents of <file> or
#			HISTORY_SAVE_FILE if none is specified.
#			Returns nothing if the read failed.
#			Returns > 0 if the read succeeds.
#		$historyctl(reset)
#			Clears the entire history buffer.
#			Returns -1 if the history buffer does not exist.
#			Returns 0 if successful.
#		$historyctl(save [<filename>])
#			Writes the current history buffer to <file> or
#			HISTORY_SAVE_FILE if none is specified.
#			Returns nothing if the write failed.
#			Returns > 0 if the write succeeds.
#		$historyctl(set <buffer size>)
#			Sets the size of the history buffer. Setting it to 0
#			purges the history buffer and disables command history.

package history;

load addset;

alias history (index, void) {
	xecho -b -c Command History:;
	if (index > (:numitems = numitems(array.history)) || !index) {
		@ index = numitems;
	};
	if (index > 0) {
		fe ($jot(0 ${index - 1} 1)) hh {
			@ :item = getitem(array.history $hh);
			xecho -c $hh${getset(history_timestamp) == 'on' ? "  \[$word(3 $stime($word(0 $item)))\]  " : "\: "}$restw(1 $item);
		};
	};
};

alias history.add (...) {
	if (@) {
		if (getset(history_remove_dupes) == 'on') {
			@ delitems(array.history $getmatches(array.history % $*));
			@ delitems(array.history $getmatches(array.history % $* ));
		};
		if (numitems(array.history) == history) {
			@ delitem(array.history 0);
		};

script/epic.py
"""High-level interface to the EPIC python integration.

This module exposes functions and classes that make interfacing your python
code with the EPIC irc client more pythonic.

Getting Started
===============

Before you can load your script the environment must be initialized. We have
provided a script for this purpose that works for the most common cases. You
can run it by typing "/load python" at the EPIC prompt. If you wish to have
the python environment initilized each time you start EPIC you can add
"load python" to your `~/.epicrc` file.

Quick Examples
==============

@alias('epic_command')
def epic_command(args):
    '''A python function working as an epic alias.
    '''
    say('Saying ' + args)
    say('Version: ' + expression('info(i)'))


@on('hook')
def on_hook(args):
    '''A python function working as an epic hook.
    '''
    say('Hooked: ' + args)


EPIC For The Python Programmer
==============================

EPIC is a fork of the ircii client <https://en.wikipedia.org/wiki/IrcII>. It
has been actively developed since 1993. It is stable and mature and offers
many benefits for both background (bot) usage and as an IRC client.

As a client, EPIC offers a robust and well tested TUI. The flexibility of the
client allows you to customize the interface in almost every way. Through the
use of script packs a user can make her client behave according to a number of
different paradigms.

As a daemon, EPIC offers a robust and well tested event framework. All you
need to worry about is what events you want to react to. EPIC handles the
differences between IRC networks and presents them to you in a coherent way.
If you run your bot inside screen or tmux you can interactively examine and
interact with your code as it's running, making debugging easier.

There are three primary ways that your python code will be executed by EPIC:

* User typed commands
* Event Triggers
* Socket Listeners

User Typed Commands
-------------------

Much like your shell, a command typed at the EPIC prompt has the full power
of the ircii language available to it. This allows the user to build powerful
constructs on their irc input line, much in the same way shells allow you to
build powerful constructs:

    > /fe (#epic #python) channel { join $channel }

You can create your own user-typed commands using the @alias() decorator
(explained below.) You can poke around the alias system yourself by typing
"/alias", or "/alias <alias_name>" at the EPIC prompt.

Event Triggers
--------------

An event trigger (aka "on" or "hook") is a piece of code that is executed
when an event happens. There are event triggers available for any event that
might happen within EPIC. Some of the more common events include connecting
to a server, joining a channel, or a message being sent to the user or a
channel the user has joined.

You can create your own event triggers using the @on() decorator (explained
below.) You can look at the existing event triggers by typing "/on" at the
EPIC prompt.

Socket Listeners
---------------

If you want to provide a network service, such as an HTTP or IRC server, you
will need to register your listening socket with EPIC. You can do this with
the `register_listener_callback()` function or the `SocketServerMixin` class,
both explained below.

Logging and Output
==================

EPIC configures a log handler for you. You don't have to worry about logging
details inside your script, simply setup a logger like so:

    import logging

    logger = logging.getLogger('my_script')
    logger.info('Hello, World!')

Anything output to STDOUT or STDERR will automatically be captured and output
into the user's EPIC window.

Classes
=======

This module comes with one class.

SocketServerMixin
-----------------

If you use a server based on Python's `socketserver.BaseRequestHandler` class
you can easily hook into EPIC's callback system by using `SocketServerMixin`.
Most of the included servers, such as `socketserver.TCPServer` and
`http.server.HTTPServer`, work with this mixin.

Under the hood `epic.SocketServerMixin` registers a socket callback for
your class and overloads `handle_request()` to integrate with EPIC's event
loop.

Decorators
==========

This module comes with two decorators.

* @alias(): Register function as a command that can be typed at the prompt
* @on(): Register function to be executed when an event happens

@alias()
--------

When a function is decorated with @alias() it will be executed when the
user types the name of your alias. Anything your function returns will be
ignored. If you wish to convey information back to the user you will need
to use a `logger`, `echo()`, `xecho()`, or `command()`.

Alias functions will be called with exactly one argument- a string
containing any arguments typed after the command.

Example:

    @alias('hello')
    def hello(args):
        xecho('Hello, %s!' % args)

When the user types "/hello Python World" it will translate to
`hello('Python World')`.

All python aliases have a small ircii shim installed. You can examine
this shim by typing "/alias hello" at the EPIC prompt.

@on()
-----

When an event happens EPIC will call any registered event handlers. Events
are thrown for basically everything that happens. By the time your client
has connected to a server, before it has even joined a channel, dozens of
events have been thrown.

You use the @on() decorator to register your event handler. An event handler
is a function that takes a single argument. When called it will be passed a
string containing one or more words corresponding to the event. For now
parsing that string is up to each event handler.

Example:

    @on('privmsg', '*enemy ships*', NOISE_QUIET)
    def enemy_ships(args):
        xecho("It's a trap!")

In this example anytime someone sends a message with the phrase "enemy ships"
their client will print out a warning:

    *** It's a trap!

All python event triggers have a small ircii shim installed. You can examine
the shim installed for our example by typing "/on privmsg" at the EPIC prompt.

Calling EPIC From Python
========================

If you want EPIC to do something for you there are 3 primary ways to do so.

xecho()
-------

You can use xecho() to output text to one or more of EPIC's windows. Your
text is prepended with the default banner by default, but this can be
overridden. There are a lot of options that control where (or even if) your
text will be output. See the docstring for xecho() for more details.

Example:

    xecho("It's a trap!")

command() and evaluate()
------------------------

If you want to execute a command as if the user had typed it at the EPIC
prompt you can use either command() or evaluate(). The primary difference
between the two is whether EPIC will parse the line and expand any variables
or ircii functions it contains.

Most of the time you want to use command(), especially if you are passing in
untrusted input. This will not do any parsing of the command line.

If you use evaluate() EPIC will parse the line first to replace $-strings
with their expandos.

These functions will always return None.

expression()
------------

When you need to get a piece of data from EPIC you can use expression().
EPIC expressions are roughly analogous to python expressions. Within
an expression you don't use $ to indicate an expando.

Example, check to see if they have logging turned on:

    expression('LOG') == 'ON'

Example, get the channel operators for a channel:

    chanops = expression('chops(#epic)')

Example, get the list of servers you're connected to:

    servers = expression('myservers()')

register_listener_callback()
----------------------------

Due to the way EPIC embeds python your python code will not have its own main
loop. To allow network services to work EPIC lets you register a callback for
a network socket using `register_listener_callback()`. When new connections
come in your callback will be called so you can handle the incoming request.

Calling Python from EPIC
========================

There are two primary ways to call Python from within EPIC. You can execute
a statement using /python or execute an expression you using $python().

/python
-------

You can use /python to execute a python statement from the top-level
(__main__) namespace. You can not execute an expression using /python,
if you attempt to do so a stack trace will be output to the currently
active window.

$python()
---------

You can use $python() to execute a python expression and get the resulting
object back. Strings will be passed as bare strings, any other object
type will be passed back as a repr() string.

Helpful Aliases
---------------

There are a couple helpful aliases you can use when developing.

* pyecho: Calls $python() and echos the output to your window
* pyload: Imports a python module
* pyreload: Re-imports a python module so you don't have to quit the client

Everything Else
===============

There are a lot of nooks and crannies that are outside the scope of this
document. It is hoped that this gives you a good grounding for writing
python code that interfaces with EPIC, but it does not talk about how
to actually use EPIC. EPIC has a rich feature set which you can access
from Python, please take some time to browse the full documentation
here:

    http://epicsol.org/help_root

If you want to know more, or just want to chat with like-minded programmers,
please come talk to us:

    Network: EFnet (irc.efnet.net)
    Channel: #epic

Acknowledgements
================

EPIC5 and the low-level Python functionality was written by hop.

This document and the high-level Python module was written by skullY.

A big thanks to hop for being willing to add "python support that doesn't
suck". He has spent many hours both talking to me and learning about how
to embed python in his program.

Finally, thanks to caf and everyone else in #epic for providing feedback
along the way. This integration is stronger for it.
"""


# FIXME: Remove this
NOTES = """
hop said he should implement _epic.set_set().
It exists, but it throws a "not implemented" exception.
You can use symbolctl(PMATCH BUILTIN_VARIABLE name) will return an empty string if that set does't exist.
You can PMATCH, CREATE, and DELETE.
This is the only way to get at some things like builtin variables.
"""

import logging
import sys
from importlib import reload

from _epic import callback_when_readable, cancel_callback, cmd, eval, expand, expr, echo, say, call
from _epic import run_command, call_function, get_set, get_assign, get_var, builtin_cmd

# Map some commands to friendlier names
command = cmd
evaluate = eval
expression = expr

# Store some potentially useful data
EPIC_BINARY_CKSUM = expression('info(s)')
EPIC_COMMIT_ID = expression('info(i)')
EPIC_COMPILE_INFO = expression('info(c)')
EPIC_COMPILE_OPTS = expression('info(o)')
EPIC_NEW_MATH_PARSER = expression('info(m)') == '1'
EPIC_RELEASE_NAME = expression('info(r)')
EPIC_RELEASE_VERSION = expression('info(v)').split(' ', 1)[1].replace(' ', '.')

# Noise flags
NOISE_DEFAULT = ''
NOISE_SILENT = '^'
NOISE_QUIET = '-'
NOISE_NOISY = '+'
NOISE_SYSTEM = '%'

# Tracking object for registered socket listeners.
# Format: _listening_sockets[<fd_num>] = (<dispatch_function>, <cleanup_function>)
_listening_sockets = {}


# Classes that help script writers interact with epic
class Console(object):
    """Write a file-like object to the epic console.
    """
    def __init__(self, output_name='PYTHON-CONSOLE'):
        self.output_name = output_name

    def write(self, buf):
        for line in buf.rstrip().splitlines():
            if self.output_name:
                echo("%s: %s" % (self.output_name, line))
            else:
                echo(line)

    def flush(self):
        pass


class SocketServerMixin(object):
    """Modify most SocketServers to work with EPIC.

    Because EPIC is a single-threaded program that controls the main loop,
    Python code is only run when EPIC calls into it. This mixin overrides
    `handle_request()` so that EPIC can dispatch to that function when a new
    connection happens. The listening socket is automatically registered with
    EPIC's callback system.

    Set `self.server_name` to change the name reported in log messages.
    """
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.server_name = self.__class__.__name__
        register_listener_callback(self.socket.fileno(), self.handle_request, self.server_close)

    def handle_request(self, vfd):
        """Handle a single request from the parent socket.
        """
        self._handle_request_noblock()
        self.service_actions()

    def server_close(self, vfd=None, *args, **kwargs):
        """Remove the callback and close the parent socket.
        """
        cancel_callback(self.socket.fileno())
        super().server_close(*args, **kwargs)


# Configure our output
sys.stdout = Console('PYTHON-STDOUT')
sys.stderr = Console('PYTHON-STDERR')
log = logging.getLogger('epic')
log.setLevel(logging.DEBUG)
log_handler = logging.StreamHandler(Console(None))
log_handler.setLevel(logging.INFO)
#log_formatter = logging.Formatter('%(asctime)s: %(name)s: %(levelname)s: %(message)s', '%H:%M:%S')
log_formatter = logging.Formatter('%(asctime)s: python.%(module)s: %(message)s', '%H:%M:%S')
log_handler.setFormatter(log_formatter)
log.addHandler(log_handler)


# Functions that epic python scripts can utilize.
def set_set(set_key, set_value, *, quiet=True):
    """Assign a value to a /set in EPIC.

    A /set is a key-value pair that always exists.
    """
    quiet = '^' if quiet else ''

    return command('%sSET %s %s' % (quiet, set_key, set_value))


def assign(assign_key, assign_value, *, quiet=True):
    """Assign a value to a /assign in EPIC.

    A /assign is a key-value pair that may or may not exist.
    """
    quiet = '^' if quiet else ''

    return command('%sASSIGN %s %s' % (quiet, set_key, set_value))


def xecho(message, all=False, all_server=False, banner=True, current=False,
    e=None, f=False, level=None, line=None, nolog=False, raw=False, say=False,
    target=None, visible=False, window=None, x=False):
    """Output a line to the user's window.

    `message`
        The message to output

    `all`
        Output to all windows.

    `all_server`
        Output to all windows on the server.

    `banner`
        Prefix the message with the current banner

    `current`
        Output to current server's current window. This is different from the
        visible option.

    `e`
        After specified number of seconds, echoed line is erased.

    `f`
        Do not notify (%F) if the window is hidden.

    `level`
        Use the given level to hunt for a window, overriding the current level
        being used.

    `line`
        When used with `win` overwrite a particular line in that window.

    `nolog`
        Do not save the message to any log files that might apply.

    `raw`
        Output the message as a raw string to the underlying tty. This is used
        to send control characters to the terminal emulator.

    `say`
        Do not output if display is being suppressed.

    `target`
        Use the given target to hunt for a window, overriding the current
        target being used.

    `visible`
        Output to a visible window. Usually the current window or the top
        window on the main screen.

    `window`
        Output to the specified window.

    `x`
        Overrule `set mangle_display` and pretend it was set to NORMALIZE
        instead.
    """
    xecho_args = ['xecho']

    if all:
        xecho_args.append('-all')
    if all_server:
        xecho_args.append('-as')
    if banner:
        xecho_args.append('-b')
    if current:
        xecho_args.append('-current')
    if isinstance(e, int):
        xecho_args.append('-e %d' % e)
    if f:
        xecho_args.append('-f')
    if isinstance(level, int):
        xecho_args.append('-level %d' % level)
    if nolog:
        xecho_args.append('-nolog')
    if raw:
        xecho_args.append('-raw')
    if say:
        xecho_args.append('-say')
    if isinstance(target, int):
        xecho_args.append('-target %d' % target)
    if visible:
        xecho_args.append('-visible')
    if isinstance(window, int):
        xecho_args.append('-window %d' % window)
        if isinstance(line, int):
            xecho_args.append('-line %d' % line)
    if x:
        xecho_args.append('-x')

    # Print the message
    xecho_args.append('--')
    xecho_args.append(message)
    command(' '.join(xecho_args))


# Decorators for registering python functions as aliases or hooks
def alias(name):
    """A decorator used to register an epic alias.

    Epic aliases will always be called with a single argument, and that
    argument will be a string. Aliases are responsible for doing their own
    argument parsing.
    """
    def decorator(f):
        module = f.__module__
        function = f.__name__
        command("alias %s {pydirect %s.%s $*}" % (name, module, function))
        expression("symbolctl(SET %s 0 ALIAS PACKAGE %s.py)" % (function,
            module))
        return f

    return decorator


def on(event_type, wildcard_pattern='*', noise_indicator=NOISE_SILENT,
    exclude_match=False, delete=False, serial_number='-',
    flexible_pattern=False):
    """A decorator used to register an epic event handler.

    For complete detail about how epic event handlers work consult the epic5
    documentation:

        <http://epicsol.org/on>

    Uniqueness
    ----------

    ONs have to be “unique”. The primary key of an ON is:

        * Event Type
        * Serial Number
        * Wildcard Pattern

    That means every ON has to have these three pieces of information.
    Creating an ON is like an “upsert” – if there is not an existing ON, you
    will create it. If there is an existing ON, you will update it. If you
    do not provide a serial number one will be selected for you.

    Serial Numbers
    --------------

    Every ON has a serial number. The default is '-'. You will want to choose
    a unique serial number for your script or leave the default, which will
    automatically choose an unused serial number less than 0.

    For each serial number at most one `@on()` will be executed. EPIC will
    determine the best match among the available Wildcard Patterns and execute
    only that hook.

    Serial numbers are executed in order. Serial number 0 is special in that
    it can supress EPIC's default output for a hook, see the full documentation
    for /on for more detail: <http://epicsol.org/on>

    Arguments
    ---------

    `event_type`
        The event that we're hooking, EG: PUBLIC or MSGS. The full list of
        possible event types is on <http://epicsol.org/help_root>, and can
        by found by looking for the pages that start with `on_`.

    `wildcard_pattern`
        The pattern you want to match. Within a pattern * matches any
        text while % matches any text until a space is encountered.

    `exclude_match`
        When True the `wildcard_pattern` is treated as an exclusion rather
        than a match.

    `delete`
        When True delete any matching hooks.

    `serial_number`
        The serial number for this hook. Use '-' to select the next unused
        number below 0, '+' to select the next unused number above 0, or
        specify your own integer here.

    `noise_indicator`
        A flag indicating how noisy this hook should be. By default all hooks
        will output all echoed messages plus an alert, "ON <event_type> hooked
        by ...".

        * NOISE_DEFAULT: Show the echoed output plus the alert.
        * NOISE_SILENT: suppress all output, do not run the *default action*.
        * NOISE_QUIET: suppress all output, run the *default action*.
        * NOISE_NOISY: show all output
        * NOISE_SYSTEM: display echoed output and supress the *default action*

    `flexible_pattern`
        When true the `wildcard_pattern` will be expanded every time the hook
        is matched. When false the `wildcard_pattern` will be matched as is.
    """
    if serial_number:
        sni = '#'
        serial_number = ' ' + str(serial_number)
    else:
        sni = ''
        serial_number = ''

    delete = '-' if delete else ''
    exclude_match = '!' if exclude_match else ''
    quote_type = "'" if flexible_pattern else '"'

    if noise_indicator not in (NOISE_DEFAULT, NOISE_SILENT, NOISE_QUIET,
                              NOISE_NOISY, NOISE_SYSTEM):
        echo('Unknown noise_indicator %s' % noise_indicator)
        noise_indicator = NOISE_DEFAULT

    def decorator(f):
        log.debug("on %s%s%s%s %s%s%s%s%s {pydirect %s.%s $*}",
            sni, noise_indicator, event_type, serial_number, exclude_match,
            delete, quote_type, wildcard_pattern, quote_type, f.__module__,
            f.__name__
        )
        command("on %s%s%s%s %s%s%s%s%s {pydirect %s.%s $*}" % (
            sni, noise_indicator, event_type, serial_number, exclude_match,
            delete, quote_type, wildcard_pattern, quote_type, f.__module__,
            f.__name__
        ))
        return f

    return decorator


def register_listener_callback(fd, dispatch_function, cleanup_function):
    """Register a dispatch and a cleanup function for a listening file descriptor.
    """
    log.debug(
        'register_listener_callback(fd=%d, dispatch_function=%s, cleanup_function=%s)',
        fd, dispatch_function.__name__, cleanup_function.__name__
    )

    if fd in _listening_sockets:
        xecho('FD already registered: %d' % fd)
        return False

    callback_when_readable(fd, dispatch_function, cleanup_function, 0)
    _listening_sockets[fd] = (dispatch_function, cleanup_function)

    return True


@on('exit')
def cleanup_listener_callbacks(args):
    """Called by our exit hook to close up our servers.
    """
    log.debug('cleanup_listener_callbacks(args=%s)', repr(args))
    for fd, funcs in _listening_sockets.items():
        cancel_callback(fd)
        funcs[1]()  # Run the cleanup function
